build/source/engine/read_attrb.f90

```diff
@@ -128,69 +127,61 @@ USE get_ixname_module,only:get_ixAttr,get_ixType  ! access function to find inde
+ err = nf90_get_var(ncid,varid,hru_id);      if(err/=nf90_noerr)then; message=trim(message)//'problem reading hruId/'//trim(nf90_strerror(err)); return; endif
 
  ! read hru2gru_id from netcdf file
- err = nf90_inq_varid(ncid, "hru2gruId", varid); if(err/=0)then; message=trim(message)//'problem finding hru2gruId'; return; endif
- err = nf90_get_var(ncid,varid,hru2gru_id);      if(err/=0)then; message=trim(message)//'problem reading hru2gruId'; return; endif
-
- ! define the HRU and GRU indices
- gru_ix = arth(1,1,nGRU)
- hru_ix = arth(1,1,nHRU)
-
- ! create local mapping vectors
- do iGRU=1,nGRU
-  ! check
-  if(count(hru2gru_id == gru_id(iGRU))/=gru_struc(iGRU)%hruCount)then
-   write(message,'(a,1x,i0)') trim(message)//'unexpected number of HRUs for GRU', iGRU
-   err=20; return
-  endif
-  ! define indices
-  where(hru2gru_id == gru_id(iGRU))
-   hru2gru_ixGRU = gru_ix(iGRU)                       ! index of the GRU, for which a global HRU is within
-   hru2gru_ixHRU = arth(1,1,gru_struc(iGRU)%hruCount) ! indices of the local HRUs for each GRU
-  endwhere
- enddo
+ err = nf90_inq_varid(ncid, "hru2gruId", varid); if(err/=nf90_noerr)then; message=trim(message)//'problem finding hru2gruId/'//trim(nf90_strerror(err)); return; endif
+ err = nf90_get_var(ncid,varid,hru2gru_id);      if(err/=nf90_noerr)then; message=trim(message)//'problem reading hru2gruId/'//trim(nf90_strerror(err)); return; endif
 
  ! fill global mapping structures
  ! gru_struc contains the HRU index for HRUs in GRUs and the unique HRU id for each HRU in each GRU
- ! index_map contains the GRU index associated with each HRU and the global HRU index (1-nHRU)
- if(allocated(gru_struc) .and. allocated(index_map) )then
-  do iHRU=1,nHRU
-
-   ! get indices
-   ix_gru = hru2gru_ixGRU(iHRU) ! index of the GRU, for which a global HRU is within
-   ix_hru = hru2gru_ixHRU(iHRU) ! index of the local HRU within each GRU
-
-   ! GRU index and id, for each global HRU
-   index_map(iHRU)%gru_ix = ix_gru  ! GRU index for each HRU
-   index_map(iHRU)%ihru   = ix_hru  ! HRU index for each HRU
-
-   ! global HRU index and id, for HRUs in GRUs
-   gru_struc(ix_gru)%hruInfo(ix_hru)%hru_ix = hru_ix(iHRU)   ! global HRU index
-   gru_struc(ix_gru)%hruInfo(ix_hru)%hru_id = hru_id(iHRU)   ! global HRU id
-
-  enddo
- else
-  err = 40
-  if(err/=0)then; message=trim(message)//'gru_struc and/or index_map not associated'; return; endif
+ if (checkHRU == missingInteger) then   
+  ! non single-HRU run 
+  
+  ! read gru_id from netcdf file
+  err = nf90_inq_varid(ncid, "gruId", varid);                              if(err/=nf90_noerr)then; message=trim(message)//'problem finding gruId/'//trim(nf90_strerror(err)); return; endif
+  err = nf90_get_var(ncid,varid,gru_id,start=(/startGRU/),count=(/nGRU/)); if(err/=nf90_noerr)then; message=trim(message)//'problem reading gruId/'//trim(nf90_strerror(err)); return; endif
+  globalHRU=0; countHRU=0
+  do ncdfHRU=1,maxHRU
```

Comment:
@martynpclark : please see my comment on the `where` statement. If the `where` statement cannot be used, the HRU loop is unavoidable.